src/neo4j_graphrag/experimental/components/text_splitters/fixed_size_splitter.py

```diff
@@ -56,10 +111,34 @@ async def run(self, text: str) -> TextChunks:
         """
         chunks = []
         index = 0
-        for i in range(0, len(text), self.chunk_size - self.chunk_overlap):
-            start = i
-            end = min(start + self.chunk_size, len(text))
+        step = self.chunk_size - self.chunk_overlap
+        text_length = len(text)
+
+        approximate_start = 0
+        skip_adjust_chunk_start = False
+        while approximate_start < text_length:
+            if self.approximate:
+                if skip_adjust_chunk_start:
+                    start = approximate_start
+                else:
+                    start = _adjust_chunk_start(text, approximate_start)
+                # adjust start and end to avoid cutting words in the middle
+                approximate_end = min(start + self.chunk_size, text_length)
+                end = _adjust_chunk_end(text, start, approximate_end)
+                # when avoiding splitting words in the middle is not possible, revert to initial chunk end and skip adjusting next chunk start
+                if end == approximate_end:
+                    skip_adjust_chunk_start = True
+                else:
+                    skip_adjust_chunk_start = False
+            else:
+                # apply fixed size splitting with possibly words cut in half at chunk boundaries
+                start = approximate_start
+                end = min(start + self.chunk_size, text_length)
+
             chunk_text = text[start:end]
             chunks.append(TextChunk(text=chunk_text, index=index))
             index += 1
+
+            approximate_start = start + step
```

Comment:
Why not using the approximate chunk size here, so that we only have to adjust the chunk end once and restart from the same position for the next chunk?

In very simple code:

```python

start = 0
while start < text_length:
    chunk = text[start:start+chunk_length]
    if self.approximate:
        chunk = adjust_end(chunk)
    start += len(chunk) - chunk_overlap
```
?